src/QsCompiler/Transformations/SearchAndReplace.cs

```diff
@@ -243,6 +249,87 @@ private void OnExpression(TypedExpression ex)
     }
 
 
+    // routines for replacing symbols/identifiers
+
+    /// <summary>
+    /// Upon transformation, assigns each defined variable a unique name, independent on the scope, and replaces all references to it accordingly.
+    /// The original variable name can be recovered by using the static method StripUniqueName.  
+    /// This class is *not* threadsafe. 
+    /// </summary>
+    public class UniqueVariableNames
+        : ScopeTransformation<UniqueVariableNames.ReplaceDeclarations, ExpressionTransformation<UniqueVariableNames.ReplaceIdentifiers>>
+    {
+        private const string Prefix = "qsVar";
+        private const string OrigVarName = "origVarName";
+        private static Regex WrappedVarName = new Regex($"^__{Prefix}[0-9]*__(?<{OrigVarName}>.*)__$");
+
+        private int VariableNr;
+        private Dictionary<NonNullable<string>, NonNullable<string>> UniqueNames;
+
+        /// <summary>
+        /// Will overwrite the dictionary entry mapping a variable name to the corresponding unique name if the key already exists. 
+        /// </summary>
+        internal NonNullable<string> GenerateUniqueName(NonNullable<string> varName)
+        {
+            var unique = NonNullable<string>.New($"__{Prefix}{this.VariableNr++}__{varName.Value}__");
```

Comment:
"this.VariableNr++" should probably be moved onto its own line, as it seems like good practice to separate the definition of "unique" from mutating the state of "VariableNr"